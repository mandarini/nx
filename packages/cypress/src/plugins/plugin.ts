import {
  CreateDependencies,
  CreateNodes,
  CreateNodesContext,
  detectPackageManager,
  NxJsonConfiguration,
  readJsonFile,
  TargetConfiguration,
  writeJsonFile,
} from '@nx/devkit';
import { dirname, extname, join, relative } from 'path';
import { registerTsProject } from '@nx/js/src/internal';

import { getLockFileName, getRootTsConfigPath } from '@nx/js';

import { CypressExecutorOptions } from '../executors/cypress/cypress.impl';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { existsSync, readdirSync } from 'fs';
import { globWithWorkspaceContext } from 'nx/src/utils/workspace-context';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { projectGraphCacheDirectory } from 'nx/src/utils/cache-directory';
import { NX_PLUGIN_OPTIONS } from '../utils/symbols';

export interface CypressPluginOptions {
  // TODO(jack): the design of this option will need to be revisited, but we do want to respect the user's package scripts.
  usePackageScripts?: boolean;
  ciTargetName?: string;
  targetName?: string;
  componentTestingTargetName?: string;
}

const cachePath = join(projectGraphCacheDirectory, 'cypress.hash');
const targetsCache = existsSync(cachePath) ? readTargetsCache() : {};

const calculatedTargets: Record<
  string,
  Record<string, TargetConfiguration>
> = {};

function readTargetsCache(): Record<
  string,
  Record<string, TargetConfiguration<CypressExecutorOptions>>
> {
  return readJsonFile(cachePath);
}

function writeTargetsToCache(
  targets: Record<
    string,
    Record<string, TargetConfiguration<CypressExecutorOptions>>
  >
) {
  writeJsonFile(cachePath, targets);
}

export const createDependencies: CreateDependencies = () => {
  writeTargetsToCache(calculatedTargets);
  return [];
};

export const createNodes: CreateNodes<CypressPluginOptions> = [
  '**/cypress.config.{js,ts,mjs,mts,cjs,cts}',
  (configFilePath, options, context) => {
    options = normalizeOptions(options);
    options.usePackageScripts ??= true;
    const projectRoot = dirname(configFilePath);

    // Do not create a project if package.json and project.json isn't there.
    const siblingFiles = readdirSync(join(context.workspaceRoot, projectRoot));
    if (
      !siblingFiles.includes('package.json') &&
      !siblingFiles.includes('project.json')
    ) {
      return {};
    }

    const hash = calculateHashForCreateNodes(projectRoot, options, context, [
      getLockFileName(detectPackageManager(context.workspaceRoot)),
    ]);

    const targets = targetsCache[hash]
      ? targetsCache[hash]
      : buildCypressTargets(configFilePath, projectRoot, options, context);

    calculatedTargets[hash] = targets;

    return {
      projects: {
        [projectRoot]: {
          projectType: 'application',
          targets,
        },
      },
    };
  },
];

function getOutputs(
  projectRoot: string,
  cypressConfig: any,
  testingType: 'e2e' | 'component'
): string[] {
  function getOutput(path: string): string {
    if (path.startsWith('..')) {
      return join('{workspaceRoot}', join(projectRoot, path));
    } else {
      return join('{projectRoot}', path);
    }
  }

  const { screenshotsFolder, videosFolder, e2e, component } = cypressConfig;
  const outputs = [];

  if (videosFolder) {
    outputs.push(getOutput(videosFolder));
  }

  if (screenshotsFolder) {
    outputs.push(getOutput(screenshotsFolder));
  }

  switch (testingType) {
    case 'e2e': {
      if (e2e.videosFolder) {
        outputs.push(getOutput(e2e.videosFolder));
      }
      if (e2e.screenshotsFolder) {
        outputs.push(getOutput(e2e.screenshotsFolder));
      }
      break;
    }
    case 'component': {
      if (component.videosFolder) {
        outputs.push(getOutput(component.videosFolder));
      }
      if (component.screenshotsFolder) {
        outputs.push(getOutput(component.screenshotsFolder));
      }
      break;
    }
  }

  return outputs;
}

function buildCypressTargets(
  configFilePath: string,
  projectRoot: string,
  options: CypressPluginOptions,
  context: CreateNodesContext
) {
  const cypressConfig = getCypressConfig(configFilePath, context);

  const pluginPresetOptions = {
    ...cypressConfig.e2e?.[NX_PLUGIN_OPTIONS],
    ...cypressConfig.env,
    ...cypressConfig.e2e?.env,
  };

  const webServerCommands: Record<string, string> =
    pluginPresetOptions?.webServerCommands;

  const relativeConfigPath = relative(projectRoot, configFilePath);

  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<string, TargetConfiguration> = {};

  if ('e2e' in cypressConfig) {
    targets[options.targetName] = {
      command: `cypress run --config-file ${relativeConfigPath} --e2e`,
      options: { cwd: projectRoot },
      cache: true,
      inputs: getInputs(namedInputs),
      outputs: getOutputs(projectRoot, cypressConfig, 'e2e'),
    };

    if (webServerCommands?.default) {
      delete webServerCommands.default;
    }

    if (Object.keys(webServerCommands ?? {}).length > 0) {
      targets[options.targetName].configurations ??= {};
      for (const [configuration, webServerCommand] of Object.entries(
        webServerCommands ?? {}
      )) {
        targets[options.targetName].configurations[configuration] = {
          command: `cypress run --config-file ${relativeConfigPath} --e2e --env webServerCommand="${webServerCommand}"`,
        };
      }
    }

    const ciWebServerCommand: string = pluginPresetOptions?.ciWebServerCommand;
    if (ciWebServerCommand) {
      const specPatterns = Array.isArray(cypressConfig.e2e.specPattern)
        ? cypressConfig.e2e.specPattern.map((p) => join(projectRoot, p))
        : [join(projectRoot, cypressConfig.e2e.specPattern)];

      const excludeSpecPatterns: string[] = !cypressConfig.e2e
        .excludeSpecPattern
        ? cypressConfig.e2e.excludeSpecPattern
        : Array.isArray(cypressConfig.e2e.excludeSpecPattern)
        ? cypressConfig.e2e.excludeSpecPattern.map((p) => join(projectRoot, p))
        : [join(projectRoot, cypressConfig.e2e.excludeSpecPattern)];
      const specFiles = globWithWorkspaceContext(
        context.workspaceRoot,
        specPatterns,
        excludeSpecPatterns
      );

      const dependsOn: TargetConfiguration['dependsOn'] = [];
      const outputs = getOutputs(projectRoot, cypressConfig, 'e2e');
      const inputs = getInputs(namedInputs);
      for (const file of specFiles) {
        const relativeSpecFilePath = relative(projectRoot, file);
        const targetName = options.ciTargetName + '--' + relativeSpecFilePath;
        targets[targetName] = {
          outputs,
          inputs,
          cache: true,
          command: `cypress run --config-file ${relativeConfigPath} --e2e --env webServerCommand="${ciWebServerCommand}" --spec ${relativeSpecFilePath}`,
          options: {
            cwd: projectRoot,
          },
        };
        dependsOn.push({
          target: targetName,
          projects: 'self',
          params: 'forward',
        });
      }
      targets[options.ciTargetName] ??= {};

      targets[options.ciTargetName] = {
        executor: 'nx:noop',
        cache: true,
        inputs,
        outputs,
        dependsOn,
      };
    }
  }

  if ('component' in cypressConfig) {
    // This will not override the e2e target if it is the same
    targets[options.componentTestingTargetName] ??= {
      command: `cypress open --config-file ${relativeConfigPath} --component`,
      options: { cwd: projectRoot },
      cache: true,
      inputs: getInputs(namedInputs),
      outputs: getOutputs(projectRoot, cypressConfig, 'component'),
    };
  }

  return targets;
}

function getCypressConfig(
  configFilePath: string,
  context: CreateNodesContext
): any {
  const resolvedPath = join(context.workspaceRoot, configFilePath);

  let module: any;
  if (['.ts', '.mts', '.cts'].includes(extname(configFilePath))) {
    const tsConfigPath = getRootTsConfigPath();

    if (tsConfigPath) {
      const unregisterTsProject = registerTsProject(tsConfigPath);
      try {
        module = load(resolvedPath);
      } finally {
        unregisterTsProject();
      }
    } else {
      module = load(resolvedPath);
    }
  } else {
    module = load(resolvedPath);
  }
  return module.default ?? module;
}

function normalizeOptions(options: CypressPluginOptions): CypressPluginOptions {
  options ??= {};
  options.targetName ??= 'e2e';
  options.componentTestingTargetName ??= 'component-test';
  options.ciTargetName ??= 'e2e-ci';
  return options;
}

function getInputs(
  namedInputs: NxJsonConfiguration['namedInputs']
): TargetConfiguration['inputs'] {
  return [
    ...('production' in namedInputs
      ? ['default', '^production']
      : ['default', '^default']),

    {
      externalDependencies: ['cypress'],
    },
  ];
}

/**
 * Load the module after ensuring that the require cache is cleared.
 */
const packageInstallationDirectories = ['node_modules', '.yarn'];

function load(path: string): any {
  // Clear cache if the path is in the cache
  if (require.cache[path]) {
    for (const k of Object.keys(require.cache)) {
      // We don't want to clear the require cache of installed packages.
      // Clearing them can cause some issues when running Nx without the daemon
      // and may cause issues for other packages that use the module state
      // in some to store cached information.
      if (!packageInstallationDirectories.some((dir) => k.includes(dir))) {
        delete require.cache[k];
      }
    }
  }

  // Then require
  return require(path);
}
