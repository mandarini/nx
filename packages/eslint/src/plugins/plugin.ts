import {
  CreateNodes,
  CreateNodesContext,
  parseJson,
  TargetConfiguration,
} from '@nx/devkit';
import { dirname, join } from 'path';
import { existsSync, readdirSync, readFileSync } from 'fs';
import { combineGlobPatterns } from 'nx/src/utils/globs';
import {
  ESLINT_CONFIG_FILENAMES,
  findBaseEslintFile,
  isFlatConfig,
} from '../utils/config-file';

export interface EslintPluginOptions {
  // TODO(jack): the design of this option will need to be revisited, but we do want to respect the user's package scripts.
  usePackageScripts?: boolean;
  targetName?: string;
}

export const createNodes: CreateNodes<EslintPluginOptions> = [
  combineGlobPatterns(['**/project.json', '**/package.json']),
  (configFilePath, options, context) => {
    const projectRoot = dirname(configFilePath);
    options = normalizeOptions(options);
    options.usePackageScripts ??= true;

    const eslintConfigs = getEslintConfigsForProject(
      projectRoot,
      context.workspaceRoot
    );
    if (!eslintConfigs.length) {
      return {};
    }

    const result = {
      projects: {
        [projectRoot]: {
          targets: buildEslintTargets(
            eslintConfigs,
            projectRoot,
            options,
            context
          ),
        },
      },
    };
    // For root projects, the name is not inferred from root package.json, so we need to manually set it.
    // TODO(jack): We should handle this in core and remove this workaround.
    if (projectRoot === '.') {
      result.projects[projectRoot]['name'] = buildProjectName(
        projectRoot,
        context.workspaceRoot
      );
    }
    return result;
  },
];

function getEslintConfigsForProject(
  projectRoot: string,
  workspaceRoot: string
): string[] {
  const detectedConfigs = new Set<string>();
  const baseConfig = findBaseEslintFile(workspaceRoot);
  if (baseConfig) {
    detectedConfigs.add(baseConfig);
  }

  let siblingFiles = readdirSync(join(workspaceRoot, projectRoot));

  if (projectRoot === '.') {
    // If there's no src folder, it's not a standalone project
    if (!siblingFiles.includes('src')) {
      return [];
    }
    // If it's standalone but doesn't have eslint config, it's not a lintable
    const config = siblingFiles.find((f) =>
      ESLINT_CONFIG_FILENAMES.includes(f)
    );
    if (!config) {
      return [];
    }
    detectedConfigs.add(config);
    return Array.from(detectedConfigs);
  }
  while (projectRoot !== '.') {
    // if it has an eslint config it's lintable
    const config = siblingFiles.find((f) =>
      ESLINT_CONFIG_FILENAMES.includes(f)
    );
    if (config) {
      detectedConfigs.add(`${projectRoot}/${config}`);
      return Array.from(detectedConfigs);
    }
    projectRoot = dirname(projectRoot);
    siblingFiles = readdirSync(join(workspaceRoot, projectRoot));
  }
  // check whether the root has an eslint config
  const config = readdirSync(workspaceRoot).find((f) =>
    ESLINT_CONFIG_FILENAMES.includes(f)
  );
  if (config) {
    detectedConfigs.add(config);
    return Array.from(detectedConfigs);
  }
  return [];
}

function buildEslintTargets(
  eslintConfigs: string[],
  projectRoot: string,
  options: EslintPluginOptions,
  context: CreateNodesContext
) {
  const isRootProject = projectRoot === '.';

  const targets: Record<string, TargetConfiguration> = {};

  const baseTargetConfig: TargetConfiguration = {
    command: `eslint ${isRootProject ? './src' : '.'}`,
    options: {
      cwd: projectRoot,
    },
  };
  if (eslintConfigs.some((config) => isFlatConfig(config))) {
    baseTargetConfig.options.env = {
      ESLINT_USE_FLAT_CONFIG: 'true',
    };
  }

  targets[options.targetName] = {
    ...baseTargetConfig,
    cache: true,
    inputs: [
      'default',
      ...eslintConfigs.map((config) => `{workspaceRoot}/${config}`),
      '{workspaceRoot}/tools/eslint-rules/**/*',
      { externalDependencies: ['eslint'] },
    ],
    options: {
      ...baseTargetConfig.options,
    },
  };

  return targets;
}

function normalizeOptions(options: EslintPluginOptions): EslintPluginOptions {
  options ??= {};
  options.targetName ??= 'lint';
  return options;
}

function buildProjectName(projectRoot: string, workspaceRoot: string): string {
  const packageJsonPath = join(workspaceRoot, projectRoot, 'package.json');
  let name: string;
  if (existsSync(packageJsonPath)) {
    const packageJson = parseJson(readFileSync(packageJsonPath, 'utf-8'));
    name = packageJson.name;
  }
  return name ?? projectRoot;
}
