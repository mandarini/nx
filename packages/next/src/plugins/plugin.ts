import {
  CreateDependencies,
  CreateNodes,
  CreateNodesContext,
  detectPackageManager,
  NxJsonConfiguration,
  parseJson,
  readJsonFile,
  TargetConfiguration,
  writeJsonFile,
} from '@nx/devkit';
import { dirname, join } from 'path';

import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { existsSync, readdirSync, readFileSync } from 'fs';

import { projectGraphCacheDirectory } from 'nx/src/utils/cache-directory';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { PHASE_PRODUCTION_BUILD } from 'next/constants';
import { getLockFileName } from '@nx/js';

export interface NextPluginOptions {
  // TODO(jack): the design of this option will need to be revisited, but we do want to respect the user's package scripts.
  usePackageScripts?: boolean;
  buildTargetName?: string;
  devTargetName?: string;
  startTargetName?: string;
}

const cachePath = join(projectGraphCacheDirectory, 'next.hash');
const targetsCache = existsSync(cachePath) ? readTargetsCache() : {};

const calculatedTargets: Record<
  string,
  Record<string, TargetConfiguration>
> = {};

function readTargetsCache(): Record<
  string,
  Record<string, TargetConfiguration>
> {
  return readJsonFile(cachePath);
}

function writeTargetsToCache(
  targets: Record<string, Record<string, TargetConfiguration>>
) {
  writeJsonFile(cachePath, targets);
}

export const createDependencies: CreateDependencies = () => {
  writeTargetsToCache(calculatedTargets);
  return [];
};

// TODO(nicholas): Add support for .mjs files
export const createNodes: CreateNodes<NextPluginOptions> = [
  '**/next.config.{js, cjs}',
  async (configFilePath, options, context) => {
    options.usePackageScripts ??= true;
    const projectRoot = dirname(configFilePath);

    // Do not create a project if package.json and project.json isn't there.
    const siblingFiles = readdirSync(join(context.workspaceRoot, projectRoot));
    if (
      !siblingFiles.includes('package.json') &&
      !siblingFiles.includes('project.json')
    ) {
      return {};
    }

    options = normalizeOptions(options);

    const hash = calculateHashForCreateNodes(projectRoot, options, context, [
      getLockFileName(detectPackageManager(context.workspaceRoot)),
    ]);

    const targets =
      targetsCache[hash] ??
      (await buildNextTargets(configFilePath, projectRoot, options, context));

    calculatedTargets[hash] = targets;

    const result = {
      projects: {
        [projectRoot]: {
          root: projectRoot,
          targets,
        },
      },
    };
    // For root projects, the name is not inferred from root package.json, so we need to manually set it.
    // TODO(jack): We should handle this in core and remove this workaround.
    if (projectRoot === '.') {
      result.projects[projectRoot]['name'] = buildProjectName(
        projectRoot,
        context.workspaceRoot
      );
    }
    return result;
  },
];

async function buildNextTargets(
  nextConfigPath: string,
  projectRoot: string,
  options: NextPluginOptions,
  context: CreateNodesContext
) {
  const nextConfig = await getNextConfig(nextConfigPath, context);
  const nextOutputPath = await getOutputs(projectRoot, nextConfig);
  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<string, TargetConfiguration> = {};
  let targetsInferredFromPackageScripts = false;

  if (options.usePackageScripts) {
    const packageJson = parseJson(
      readFileSync(
        join(context.workspaceRoot, projectRoot, 'package.json'),
        'utf-8'
      )
    );
    const nextScriptRegexp = /(next\s+(?<cmd>\w+)|\bnext-remote-watch\b)/;

    // Need to wire up `dependsOn` for start to perform build first
    let startTargetName: string;
    let buildTargetName: string;

    if (packageJson.scripts) {
      targetsInferredFromPackageScripts = true;
      for (const [scriptName, script] of Object.entries(
        packageJson.scripts as Record<string, string>
      )) {
        // If script is `next build` then we need to configure input and outputs
        const match = nextScriptRegexp.exec(script);
        if (!match) continue;

        targets[scriptName] = {
          command: script,
          options: {
            cwd: projectRoot,
          },
        };

        if (match.groups?.cmd === 'build') {
          buildTargetName = scriptName;
          targets[scriptName].dependsOn = ['^build'];
          targets[scriptName].cache = true;
          targets[scriptName].inputs = getInputs(namedInputs);
          targets[scriptName].outputs = [
            nextOutputPath,
            `${nextOutputPath}/!(cache)`,
          ];
        } else if (match.groups?.cmd === 'start') {
          startTargetName = scriptName;
        }
      }

      if (buildTargetName && startTargetName) {
        targets[startTargetName].dependsOn = [buildTargetName];
      }
    }
  }

  if (!targetsInferredFromPackageScripts) {
    targets[options.buildTargetName] = {
      command: `next build`,
      options: {
        cwd: projectRoot,
      },
      dependsOn: ['^build'],
      cache: true,
      inputs: getInputs(namedInputs),
      outputs: [nextOutputPath, `${nextOutputPath}/!(cache)`],
    };

    targets[options.devTargetName] = {
      command: `next dev`,
      options: {
        cwd: projectRoot,
      },
    };

    targets[options.startTargetName] = {
      command: `next start`,
      options: {
        cwd: projectRoot,
      },
      dependsOn: [options.buildTargetName],
    };
  }

  return targets;
}

async function getOutputs(projectRoot, nextConfig) {
  let dir = '.next';

  if (typeof nextConfig === 'function') {
    // Works for both async and sync functions.
    const configResult = await Promise.resolve(
      nextConfig(PHASE_PRODUCTION_BUILD, { defaultConfig: {} })
    );
    if (configResult?.distDir) {
      dir = configResult?.distDir;
    }
  } else if (typeof nextConfig === 'object' && nextConfig?.distDir) {
    // If nextConfig is an object, directly use its 'distDir' property.
    dir = nextConfig.distDir;
  }
  return projectRoot === '.'
    ? `{projectRoot}/${dir}`
    : `{workspaceRoot}/${projectRoot}/${dir}`;
}

async function getNextConfig(
  configFilePath: string,
  context: CreateNodesContext
): Promise<any> {
  const resolvedPath = join(context.workspaceRoot, configFilePath);

  const module = await load(resolvedPath);
  return module.default ?? module;
}

function normalizeOptions(options: NextPluginOptions): NextPluginOptions {
  options ??= {};
  options.buildTargetName ??= 'build';
  options.devTargetName ??= 'dev';
  options.startTargetName ??= 'start';
  return options;
}

function getInputs(
  namedInputs: NxJsonConfiguration['namedInputs']
): TargetConfiguration['inputs'] {
  return [
    ...('production' in namedInputs
      ? ['default', '^production']
      : ['default', '^default']),
    {
      externalDependencies: ['next'],
    },
  ];
}

/**
 * Load the module after ensuring that the require cache is cleared.
 */
async function load(path: string): Promise<any> {
  if (path.endsWith('.js') || path.endsWith('.cjs')) {
    // Clear cache if the path is in the cache
    if (require.cache[path]) {
      for (const k of Object.keys(require.cache)) {
        delete require.cache[k];
      }
    }

    // Then require
    return require(path);
  } else {
    // TODO(nicholas): handle cache clear for ESM
    return await Function(`return import("${path}")`)();
  }
}

function buildProjectName(projectRoot: string, workspaceRoot: string): string {
  const packageJsonPath = join(workspaceRoot, projectRoot, 'package.json');
  let name: string;
  if (existsSync(packageJsonPath)) {
    const packageJson = parseJson(readFileSync(packageJsonPath, 'utf-8'));
    name = packageJson.name;
  }
  return name ?? projectRoot;
}
