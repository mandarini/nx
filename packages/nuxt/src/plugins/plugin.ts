import {
  CreateDependencies,
  CreateNodes,
  CreateNodesContext,
  detectPackageManager,
  joinPathFragments,
  readJsonFile,
  TargetConfiguration,
  workspaceRoot,
  writeJsonFile,
} from '@nx/devkit';
import { basename, dirname, isAbsolute, join, relative } from 'path';
import { projectGraphCacheDirectory } from 'nx/src/utils/cache-directory';
import { getNamedInputs } from '@nx/devkit/src/utils/get-named-inputs';
import { existsSync, readdirSync } from 'fs';
import { loadNuxtKitDynamicImport } from '../utils/executor-utils';
import { calculateHashForCreateNodes } from '@nx/devkit/src/utils/calculate-hash-for-create-nodes';
import { getLockFileName } from '@nx/js';

const cachePath = join(projectGraphCacheDirectory, 'nuxt.hash');
const targetsCache = existsSync(cachePath) ? readTargetsCache() : {};

const calculatedTargets: Record<
  string,
  Record<string, TargetConfiguration>
> = {};

function readTargetsCache(): Record<
  string,
  Record<string, TargetConfiguration>
> {
  return readJsonFile(cachePath);
}

function writeTargetsToCache(
  targets: Record<string, Record<string, TargetConfiguration>>
) {
  writeJsonFile(cachePath, targets);
}

export const createDependencies: CreateDependencies = () => {
  writeTargetsToCache(calculatedTargets);
  return [];
};

export interface NuxtPluginOptions {
  // TODO(jack): the design of this option will need to be revisited, but we do want to respect the user's package scripts.
  usePackageScripts?: boolean;
  buildTargetName?: string;
  serveTargetName?: string;
}

export const createNodes: CreateNodes<NuxtPluginOptions> = [
  '**/nuxt.config.{js,ts}',
  async (configFilePath, options, context) => {
    options.usePackageScripts ??= true;
    const projectRoot = dirname(configFilePath);
    // Do not create a project if package.json and project.json isn't there.
    const siblingFiles = readdirSync(join(context.workspaceRoot, projectRoot));
    if (
      !siblingFiles.includes('package.json') &&
      !siblingFiles.includes('project.json')
    ) {
      return {};
    }

    options = normalizeOptions(options);

    const hash = calculateHashForCreateNodes(projectRoot, options, context, [
      getLockFileName(detectPackageManager(context.workspaceRoot)),
    ]);
    const targets = targetsCache[hash]
      ? targetsCache[hash]
      : await buildNuxtTargets(configFilePath, projectRoot, options, context);

    calculatedTargets[hash] = targets;

    return {
      projects: {
        [projectRoot]: {
          root: projectRoot,
          targets,
        },
      },
    };
  },
];

async function buildNuxtTargets(
  configFilePath: string,
  projectRoot: string,
  options: NuxtPluginOptions,
  context: CreateNodesContext
) {
  const nuxtConfig: {
    buildDir: string;
  } = await getInfoFromNuxtConfig(configFilePath, context, projectRoot);

  const { buildOutputs } = getOutputs(
    nuxtConfig,

    projectRoot
  );

  const namedInputs = getNamedInputs(projectRoot, context);

  const targets: Record<string, TargetConfiguration> = {};

  targets[options.buildTargetName] = buildTarget(
    options.buildTargetName,
    namedInputs,
    buildOutputs,
    projectRoot
  );

  targets[options.serveTargetName] = serveTarget(projectRoot);

  return targets;
}

function buildTarget(
  buildTargetName: string,
  namedInputs: {
    [inputName: string]: any[];
  },
  buildOutputs: string[],
  projectRoot: string
) {
  return {
    command: `nuxi build`,
    options: { cwd: projectRoot },
    cache: true,
    dependsOn: [`^${buildTargetName}`],
    inputs: [
      ...('production' in namedInputs
        ? ['production', '^production']
        : ['default', '^default']),

      {
        externalDependencies: ['nuxi'],
      },
    ],
    outputs: buildOutputs,
  };
}

function serveTarget(projectRoot: string) {
  const targetConfig: TargetConfiguration = {
    command: `nuxi dev`,
    options: {
      cwd: projectRoot,
    },
  };

  return targetConfig;
}

async function getInfoFromNuxtConfig(
  configFilePath: string,
  context: CreateNodesContext,
  projectRoot: string
): Promise<{
  buildDir: string;
}> {
  const { loadNuxtConfig } = await loadNuxtKitDynamicImport();

  const config = await loadNuxtConfig({
    cwd: joinPathFragments(context.workspaceRoot, projectRoot),
    configFile: basename(configFilePath),
  });

  return {
    buildDir: config?.buildDir,
  };
}

function getOutputs(
  nuxtConfig: { buildDir: string },
  projectRoot: string
): {
  buildOutputs: string[];
} {
  let nuxtBuildDir = nuxtConfig?.buildDir;
  if (nuxtConfig?.buildDir && basename(nuxtConfig?.buildDir) === '.nuxt') {
    // buildDir will most probably be `../dist/my-app/.nuxt`
    // we want the "general" outputPath to be `../dist/my-app`
    nuxtBuildDir = nuxtConfig.buildDir.replace(
      basename(nuxtConfig.buildDir),
      ''
    );
  }
  const buildOutputPath =
    normalizeOutputPath(nuxtBuildDir, projectRoot) ??
    '{workspaceRoot}/dist/{projectRoot}';

  return {
    buildOutputs: [buildOutputPath],
  };
}

function normalizeOutputPath(
  outputPath: string | undefined,
  projectRoot: string
): string | undefined {
  if (!outputPath) {
    if (projectRoot === '.') {
      return `{projectRoot}/dist`;
    } else {
      return `{workspaceRoot}/dist/{projectRoot}`;
    }
  } else {
    if (isAbsolute(outputPath)) {
      return `{workspaceRoot}/${relative(workspaceRoot, outputPath)}`;
    } else {
      if (outputPath.startsWith('..')) {
        return join('{workspaceRoot}', join(projectRoot, outputPath));
      } else {
        return join('{projectRoot}', outputPath);
      }
    }
  }
}

function normalizeOptions(options: NuxtPluginOptions): NuxtPluginOptions {
  options ??= {};
  options.buildTargetName ??= 'build';
  options.serveTargetName ??= 'serve';
  return options;
}
