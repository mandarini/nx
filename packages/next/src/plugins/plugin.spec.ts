import { CreateNodesContext } from '@nx/devkit';
import type { NextConfig } from 'next';

import { createNodes } from './plugin';
import { TempFs } from '@nx/devkit/internal-testing-utils';

describe('@nx/next/plugin', () => {
  let createNodesFunction = createNodes[1];
  let context: CreateNodesContext;

  describe('root projects', () => {
    beforeEach(async () => {
      context = {
        nxJsonConfiguration: {
          namedInputs: {
            default: ['{projectRoot}/**/*'],
            production: ['!{projectRoot}/**/*.spec.ts'],
          },
        },
        workspaceRoot: '',
      };
    });

    afterEach(() => {
      jest.resetModules();
    });

    it('should create nodes', async () => {
      const nextConfigPath = 'next.config.js';
      mockNextConfig(nextConfigPath, {});
      const nodes = await createNodesFunction(
        nextConfigPath,
        {
          usePackageScripts: false,
          buildTargetName: 'build',
          devTargetName: 'dev',
          startTargetName: 'start',
        },
        context
      );

      expect(nodes).toMatchSnapshot();
    });
  });

  describe('integrated projects', () => {
    const tempFs = new TempFs('test');
    beforeEach(() => {
      context = {
        nxJsonConfiguration: {
          namedInputs: {
            default: ['{projectRoot}/**/*'],
            production: ['!{projectRoot}/**/*.spec.ts'],
          },
        },
        workspaceRoot: tempFs.tempDir,
      };

      tempFs.createFileSync(
        'my-app/project.json',
        JSON.stringify({ name: 'my-app' })
      );
      tempFs.createFileSync('my-app/next.config.js', '');
    });

    afterEach(() => {
      jest.resetModules();
    });

    it('should create nodes', async () => {
      mockNextConfig('my-app/next.config.js', {});
      const nodes = await createNodesFunction(
        'my-app/next.config.js',
        {
          usePackageScripts: false,
          buildTargetName: 'my-build',
          devTargetName: 'my-serve',
          startTargetName: 'my-start',
        },
        context
      );

      expect(nodes).toMatchSnapshot();
    });

    it('should create nodes from package scripts', async () => {
      mockNextConfig('my-app/next.config.js', {});
      tempFs.createFileSync(
        'my-app/package.json',
        JSON.stringify({
          name: 'my-app',
          scripts: {
            dev: 'next dev --turbo',
            start: 'next start',
            build: 'next build',
          },
        })
      );
      const nodes = await createNodesFunction(
        'my-app/next.config.js',
        {
          usePackageScripts: true,
        },
        context
      );

      expect(nodes).toMatchSnapshot();
    });
  });
});

function mockNextConfig(path: string, config: NextConfig) {
  jest.mock(
    path,
    () => ({
      default: config,
    }),
    {
      virtual: true,
    }
  );
}
