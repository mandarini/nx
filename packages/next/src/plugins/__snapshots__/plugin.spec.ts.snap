// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nx/next/plugin integrated projects should create nodes 1`] = `
{
  "projects": {
    "my-app": {
      "root": "my-app",
      "targets": {
        "my-build": {
          "cache": true,
          "command": "next build",
          "dependsOn": [
            "^build",
          ],
          "inputs": [
            "default",
            "^production",
            {
              "externalDependencies": [
                "next",
              ],
            },
          ],
          "options": {
            "cwd": "my-app",
          },
          "outputs": [
            "{workspaceRoot}/my-app/.next",
            "{workspaceRoot}/my-app/.next/!(cache)",
          ],
        },
        "my-serve": {
          "command": "next dev",
          "options": {
            "cwd": "my-app",
          },
        },
        "my-start": {
          "command": "next start",
          "dependsOn": [
            "my-build",
          ],
          "options": {
            "cwd": "my-app",
          },
        },
      },
    },
  },
}
`;

exports[`@nx/next/plugin integrated projects should create nodes from package scripts 1`] = `
{
  "projects": {
    "my-app": {
      "root": "my-app",
      "targets": {
        "build": {
          "cache": true,
          "command": "next build",
          "dependsOn": [
            "^build",
          ],
          "inputs": [
            "default",
            "^production",
            {
              "externalDependencies": [
                "next",
              ],
            },
          ],
          "options": {
            "cwd": "my-app",
          },
          "outputs": [
            "{workspaceRoot}/my-app/.next",
            "{workspaceRoot}/my-app/.next/!(cache)",
          ],
        },
        "dev": {
          "command": "next dev --turbo",
          "options": {
            "cwd": "my-app",
          },
        },
        "start": {
          "command": "next start",
          "dependsOn": [
            "build",
          ],
          "options": {
            "cwd": "my-app",
          },
        },
      },
    },
  },
}
`;

exports[`@nx/next/plugin root projects should create nodes 1`] = `
{
  "projects": {
    ".": {
      "name": "@nx/nx-source",
      "root": ".",
      "targets": {
        "build": {
          "cache": true,
          "command": "next build",
          "dependsOn": [
            "^build",
          ],
          "inputs": [
            "default",
            "^production",
            {
              "externalDependencies": [
                "next",
              ],
            },
          ],
          "options": {
            "cwd": ".",
          },
          "outputs": [
            "{projectRoot}/.next",
            "{projectRoot}/.next/!(cache)",
          ],
        },
        "dev": {
          "command": "next dev",
          "options": {
            "cwd": ".",
          },
        },
        "start": {
          "command": "next start",
          "dependsOn": [
            "build",
          ],
          "options": {
            "cwd": ".",
          },
        },
      },
    },
  },
}
`;
